src/fswatch_win/fswatch_win_stubs.c

```diff
@@ -0,0 +1,430 @@
+  DWORD name_len;
+  WCHAR* file_name;
+
+  while (events != NULL) {
+    info = events->buffer;
+    v_directory = caml_copy_string(events->dirpath);
+
+    for (;;) {
+      name_len = info->FileNameLength / sizeof(WCHAR);
+      file_name = caml_stat_alloc(sizeof(WCHAR) * (name_len + 1));
+      memcpy(file_name, info->FileName, sizeof(WCHAR) * name_len);
+      file_name[name_len] = 0;
+      v_path = caml_copy_string_of_utf16(file_name);
+      caml_stat_free(file_name);
+
+      /* Allocate Event.t */
+      ev = caml_alloc_tuple(3);
+      Store_field(ev, 0, v_directory);
+      Store_field(ev, 1, v_path);
+      Store_field(ev, 2, Val_action(info->Action));
+
+      /* Allocate list cell */
+      cell = caml_alloc_tuple(2);
+      Store_field(cell, 0, ev);
+      Store_field(cell, 1, res);
+      res = cell;
+
+      if (info->NextEntryOffset == 0)
+        break;
+
+      info = (PFILE_NOTIFY_INFORMATION)((char*)info + info->NextEntryOffset);
+    }
+
+    next = events->next;
+    free(events->buffer);
+    free(events->dirpath);
+    free(events);
+    events = next;
+  }
+
+  CAMLreturn(res);
+}
+
+static value get_events(struct fsenv *fsenv) {
+  struct events *events = pop_events(fsenv);
+
+  if (ResetEvent(fsenv->signal) == 0)
+    unix_error("ResetEvent", GetLastError());
+
+  return parse_events(events);
+}
+
+value fswatch_win_wait(value v_fsenv, value v_debounce) {
+  struct fsenv *fsenv = (struct fsenv *)Nativeint_val(v_fsenv);
+  DWORD res;
+
+  caml_release_runtime_system();
+  res = WaitForSingleObject(fsenv->signal, INFINITE);
+  if (res == WAIT_OBJECT_0 && Int_val(v_debounce) > 0) Sleep(Int_val(v_debounce));
+  caml_acquire_runtime_system();
+
+  if (res != WAIT_OBJECT_0)
+    unix_error("WaitForSingleObject", GetLastError());
+
+  return get_events(fsenv);
+}
+
+#define FILE_NOTIFY_CONDITIONS \
+  (FILE_NOTIFY_CHANGE_FILE_NAME | FILE_NOTIFY_CHANGE_DIR_NAME | FILE_NOTIFY_CHANGE_LAST_WRITE)
+
+static DWORD read_changes(struct wenv *wenv) {
+  memset(wenv->overlapped, 0, sizeof(OVERLAPPED));
+  if (wenv->buffer == NULL) wenv->buffer = malloc(FILE_NOTIFY_BUFFER_LENGTH);
+  return
+    ReadDirectoryChangesW(wenv->dir_handle, wenv->buffer, FILE_NOTIFY_BUFFER_LENGTH, TRUE,
+                          FILE_NOTIFY_CONDITIONS, NULL, wenv->overlapped, NULL);
+}
+
+static WCHAR* utf8_to_utf16(const char* s, int len) {
+  WCHAR* buf;
+  int buflen = MultiByteToWideChar(CP_UTF8, 0, s, len, NULL, 0);
+  if (buflen == 0) {
+    DEBUG("MultiByteToWideChar: %s: error %d", s, GetLastError());
+    return NULL;
+  }
+  buf = malloc(buflen * sizeof(WCHAR));
+  if (MultiByteToWideChar(CP_UTF8, 0, s, len, buf, buflen) == 0) {
+    DEBUG("MultiByteToWideChar: %s: error %d", s, GetLastError());
+    free(buf);
+    return NULL;
+  }
+  return buf;
+}
+
+static void free_wenv(struct wenv *wenv, BOOL closeHandle) {
+  free(wenv->dirpath);
+  free(wenv->realpath);
+  free(wenv->overlapped);
+  if (wenv->buffer) free(wenv->buffer);
```

Comment:
```suggestion
  free(wenv->buffer);
```
No need to check that it's not `NULL`, `free` does it.